doc/event-alarm-framework/events-producer.md

```diff
@@ -0,0 +1,212 @@
+- A new event addition could be as simple as copying couple of small files to a switch and a rsyslog/container restart.
+
+##### Design at high level
+- Configure a rsyslog plugin as per process with rsyslog.d.
+- For logs raised by host processes, configure this plugin at host.
+- For logs raised by processes inside the container, configure for rsyslog.d running inside the container.
+- In this mode, there will be plugin processes running in host & in containers as one instance per process.
+- Provide the regex patterns to use for matching events as i/p to the plugin (*list of patterns for a process*).
+- Each plugin scans messasges **only** from a *single* process and matches only against patterns for that process.
+- For messages that match a pattern, retrieve parameters of interest and fire event using event reporter.
+- The rsyslog plugin binary, which does the parsing & reporting is a single binary in host, shared/used by all plugins.
+- The rsyslog plugin binary being under host control, ensures a single/unified behavior across all.
+- The unit tests can use hardcoded log messages to validate regex.
+
+
+![image](https://user-images.githubusercontent.com/47282725/157343412-6c4a6519-c27b-459b-896b-7875d8f952b8.png)
+
+##### Pros & Cons
+
+###### Pro
+1) This support is external to app, hence adapts well with III party applications, bgp, teamd, ... 
+2) This feature can be added to released builds too, as all it takes is to copy two files into each container and restart rsyslogd in the container
+3) The regex for parsing being local to container, it supports any container upgrade transaparently.
+4) The rsyslog plugin binary is maintained by host, hence provide a overall control across all containers.
+5) The message parsing load is distributed as per container. Within a container parsing is done at the granularity of per process with no extra cost as rsyslogd already pre-parsed it per-process, *always*
+
+###### con:
+1) Two step process for devs, as for each new/updated log message ***for an event***, add/update regex as needed. Updating code directly to raise the event will help avoid this.
+2) The rsyslogd is *required*. It should be treated as critical process in each container.
+
+### Event reporting
+
+#### requirements
+- Events are reported from multiple event detectors or we may call event-sources.
+- The event detectors could be running in host and/or some/all containers.
+- Support multiple local clients to be able to receive the events concurrently.
+- Each local client should be able to receive updates from all event detectors w/o being aware of all of the sources.
+- The local clients could be operating at different speeds.
+- The clients may come and go.
+- There may not be any local client to receive the events.
+- Events reporting should not be blocked by any back pressure from any local client.
+
+#### Design
+- Event detectors send UDP messages to a multicast group
```

Comment:
>multicast group [](http://example.com/codeflow?start=41&length=15)

Could you explain "send UDP messages to a multicast group" in details?

Reply:
many to many. -- UDP multicast concept.
Again, if the telemetry client is designed to accept UDP stream.